doc/sphinx/proof-engine/tactics.rst

```diff
@@ -142,43 +142,41 @@ introduced by tactics.  They also let you split an introduced hypothesis into
 multiple hypotheses or subgoals.  Common tactics that accept intro patterns
 include :tacn:`assert`, :tacn:`intros` and :tacn:`destruct`.
 
+.. insertprodn intropattern or_and_intropattern
+
 .. prodn::
-   intropattern_list ::= {* @intropattern }
    intropattern ::= *
    | **
    | @simple_intropattern
    simple_intropattern ::= @simple_intropattern_closed {* % @term0 }
-   simple_intropattern_closed ::= @naming_intropattern
+   simple_intropattern_closed ::= @or_and_intropattern
+   | @equality_intropattern
    | _
-   | @or_and_intropattern
-   | @rewriting_intropattern
-   | @injection_intropattern
-   naming_intropattern ::= @ident
+   | @naming_intropattern
+   equality_intropattern ::= ->
+   | <-
+   | [= {* @intropattern } ]
```

Comment:
Why did you splice `injection_intropattern` and `rewriting_intropattern`? These were useful to distinguish.

Reply:
They're not spliced and not my handiwork.  Those nonterminals don't appear in the mlg files.  They didn't exist as of cf868740c3d18ee9ce9a6b38dd617784625a3cae, the commit that added doc_grammar (dated 1/2/2019).

The doc is a wee bit out of date--who would have guessed?

We could add some text if there are semantic restrictions.

Reply:
> The doc is a wee bit out of date--who would have guessed?

This is a part of the documentation that we've already worked on together (more than two years ago, see #9269). So I'm a bit disappointing to discover that this grammar is changed. If you can introduce edits that would make it more like what is currently in the manual, without sacrificing correctness, this would ease my mind.

Reply:
Addressed [here](https://github.com/coq/coq/pull/14179#discussion_r653332728)

Reply:
I don't see how this is addressed here.

Reply:
Well, you asked why, which is what I responded to rather than to your implicit request to put it back as it was.  I'll put it back.

Reply:
> So I'm a bit disappointing to discover that this grammar is changed.

The nonterminals were added by Hugo without anything in the mlgs or `common.edit_mlg`.  So easy to drop them without much thought.

Reply:
You believe `injection` and `einjection` only take `injection_intropattern`?

The mlgs have:

```
| "injection" "as" LIST0 simple_intropattern
| "injection" destruction_arg "as" LIST0 simple_intropattern
| "einjection" "as" LIST0 simple_intropattern
| "einjection" destruction_arg "as" LIST0 simple_intropattern
```

You're sure these tactics are semantically restricted to _only_ accept `"[=" LIST0 intropattern "]"`?  If so, then the mlg grammar for these commands should be simplified.

```
simple_intropattern: [
| simple_intropattern_closed LIST0 [ "%" term0 ]
]

simple_intropattern_closed: [
| or_and_intropattern
| equality_intropattern
| "_"
| naming_intropattern
]

equality_intropattern: [
| "->"
| "<-"
| "[=" LIST0 intropattern "]"
]

naming_intropattern: [
| "?" ident
| "?"
| ident
]

or_and_intropattern: [
| "[" LIST0 ( LIST0 intropattern ) SEP "|" "]"
| "(" LIST0 simple_intropattern SEP "," ")"
| "(" simple_intropattern "&" LIST1 simple_intropattern SEP "&" ")"
]
```

Reply:
Are there any tactics that should use `rewriting_intropattern` in their syntax?  It sure doesn't make sense for `rewrite`.  The nonterminal doesn't appear anywhere else in the rst files other than in the prodns here, so how exactly is it valuable?

Reply:
`injection` indeed also accepts a simple list of intropatterns IIRC, and you are right that rewriting intropatterns are not used specifically by any tactic. But gathering them inside semantic units makes it easier to understand the various categories IMHO.